### DISCUSSION TITLE
2.1 ループを使って動的にプロパティ名を生成したい場合は?

### DISCUSSION THREAD

**Discussion by junseinagao** [tricot-inc/frontend-style-guide#4] [Q&A] [answered]
# 質問

> [2.1](https://github.com/tricot-inc/frontend-style-guide#2-1) 動的にプロパティ名を持つオブジェクトを作成する場合もオブジェクトの中で定義を行いましょう。

において動的にプロパティ名を生成したい場合は、複数のプロパティをループで生成したいケースもあると思うのですが、このスタイルガイドではそれを非推奨していると思います。このような場合はどのように実装されてますか?

もしよろしければ、スタイルガイドをより良くするためにもこの個人の疑問に答えてくださると嬉しいです!

```js
function getKey(key: string): string {
  return `key-is-${key}`;
}

const keys = ['Keys', 'that', 'you', 'must' 'add', 'to', 'object', 'and', 'this', 'elements', 'are', 'changed', 'dynamic']

// bad
const obj = {
  id: 1,
};
for (let i = 0;i < keys.length;i++) {
  obj[getKey(keys[i])] = true
}

// good 
const obj = {
  id: 1,
  // How ?
};
```

**Comment by konojunya** [comment, marked as answer]
ありがとうございます。
まずそもそもこの配列が事前決定されている場合のみ `obj` の定義の際に一緒に作れると思いますが、もしそのエッジケースで全ての value に `true` を入れたい場合は以下のように実装ができると思います。

```js
const obj = {
    id: 1,
    ...keys.reduce((pre, cur) => ({...pre, [getKey(cur)]: true}), {})
}
```

**Reply by junseinagao** [reply]
@konojunya 

ありがとうございます！🙇‍♂️

配列の束をオブジェクトに変換してから分割代入すれば良いのは見落としてました！素晴らしいと思います！
